dask/base.py

```diff
@@ -60,3 +62,27 @@ def compute(*args, **kwargs):
     keys = [arg._keys() for arg in args]
     results = get(dsk, keys, **kwargs)
     return tuple(a._finalize(a, r) for a, r in zip(args, results))
+
+
+normalize = Dispatch()
+normalize.register(object, lambda a: a)
+normalize.register(dict, lambda a: tuple(sorted(a.items())))
+with ignoring(ImportError):
+    import pandas as pd
+    normalize.register(pd.DataFrame, lambda a: (id(a), len(a), list(a.columns)))
+    normalize.register(pd.Series, lambda a: (id(a), len(a), a.name))
+with ignoring(ImportError):
+    import numpy as np
+    normalize.register(np.ndarray, lambda a: (id(a), a.dtype, a.shape))
```

Comment:
Yeah, this is a tricky topic I suppose.  I would probably do this differently but you're the one doing the work.  I don't personally object to this.  

I might do something hacky like the following.

def normalize(o):
    if isinstance(o, (str, bytes, unicode):
        ...
    <lots of other things that are always around
    name = type(o).__module + '.' + type(o).__name__
    if name == 'numpy.ndarray':
        return id(o), o.shape, o.dtype
    if name == 'pandas.DataFrame':
        ...
```

I like this because it's low tech and also fairly simple.  It's not extensible without touching the code but we haven't run into that case yet.

That being said, the above example is only an example of my thinking.  If you prefer the `Dispatch` object then go for it.  These things are also easy to change in the future if necessary.